Index: qmiproxy-d362b032ac8aaf8831afea840a07628acfe9b01f-2020-03-26/Makefile
===================================================================
--- qmiproxy-d362b032ac8aaf8831afea840a07628acfe9b01f-2020-03-26.orig/Makefile
+++ qmiproxy-d362b032ac8aaf8831afea840a07628acfe9b01f-2020-03-26/Makefile
@@ -1,5 +1,5 @@
-CFLAGS=-std=c99 -Wall -Wextra -Werror
-LDFLAGS=-lutil -lpthread
+CFLAGS+=-std=c99 -Wall -Wextra -Werror
+LDFLAGS+=-lutil -lpthread
 
 all: qmiserial2qmuxd
 
