--- a/drivers/net/ethernet/qualcomm/emac/emac.c
+++ b/drivers/net/ethernet/qualcomm/emac/emac.c
@@ -15,6 +15,8 @@
 #include <linux/phy.h>
 #include <linux/platform_device.h>
 #include <linux/acpi.h>
+#include <linux/regulator/consumer.h>
+#include <linux/pinctrl/consumer.h>
 #include "emac.h"
 #include "emac-mac.h"
 #include "emac-phy.h"
@@ -64,12 +66,21 @@
 	DMAW_TO_INT     |\
 	TXQ_TO_INT)
 
+#define EMAC_PINCTRL_STATE_MDIO_ACTIVE "emac_mdio_active"
+#define EMAC_PINCTRL_STATE_MDIO_SLEEP  "emac_mdio_sleep"
+#define EMAC_PINCTRL_STATE_EPHY_ACTIVE "emac_ephy_active"
+#define EMAC_PINCTRL_STATE_EPHY_SLEEP  "emac_ephy_sleep"
+
 /* in sync with enum emac_clk_id */
 static const char * const emac_clk_name[] = {
 	"axi_clk", "cfg_ahb_clk", "high_speed_clk", "mdio_clk", "tx_clk",
 	"rx_clk", "sys_clk"
 };
 
+static const char * const emac_regulator_name[] = {
+	"emac_vreg1", "emac_vreg2", "emac_vreg3", "emac_vreg4", "emac_vreg5"
+};
+
 void emac_reg_update32(void __iomem *addr, u32 mask, u32 val)
 {
 	u32 data = readl(addr);
@@ -550,6 +561,103 @@ static void emac_clks_teardown(struct em
 		clk_disable_unprepare(adpt->clk[i]);
 }
 
+static inline int msm_emac_request_pinctrl_on(struct emac_adapter *adpt,
+					      bool mdio, bool ephy)
+{
+	int result = 0;
+	int ret    = 0;
+
+	if (mdio) {
+		result = pinctrl_select_state(adpt->pinctrl,
+					      adpt->mdio_pins_active);
+		if (result)
+			netdev_err(adpt->netdev, "error:%d Can not switch on %s pins\n",
+					result, EMAC_PINCTRL_STATE_MDIO_ACTIVE);
+		ret = result;
+	}
+
+	if (ephy) {
+		result = pinctrl_select_state(adpt->pinctrl,
+					      adpt->ephy_pins_active);
+		if (result)
+			netdev_err(adpt->netdev, "error:%d Can not switch on %s pins\n",
+					result, EMAC_PINCTRL_STATE_EPHY_ACTIVE);
+		if (!ret)
+			ret = result;
+	}
+	return ret;
+}
+
+static inline int msm_emac_request_pinctrl_off(struct emac_adapter *adpt,
+					       bool mdio, bool ephy)
+{
+	int result = 0;
+	int ret    = 0;
+
+	if (mdio) {
+		result = pinctrl_select_state(adpt->pinctrl,
+					      adpt->mdio_pins_sleep);
+		if (result)
+			netdev_err(adpt->netdev, "error:%d Can not switch off %s pins\n",
+					result, EMAC_PINCTRL_STATE_MDIO_SLEEP);
+		ret = result;
+	}
+
+	if (ephy) {
+		result = pinctrl_select_state(adpt->pinctrl,
+					      adpt->ephy_pins_sleep);
+		if (result)
+			netdev_err(adpt->netdev, "error:%d Can not switch off %s pins\n",
+					result, EMAC_PINCTRL_STATE_EPHY_SLEEP);
+		if (!ret)
+			ret = result;
+	}
+	return ret;
+}
+
+static int msm_emac_pinctrl_init(struct emac_adapter *adpt, struct device *dev)
+{
+	adpt->pinctrl = devm_pinctrl_get(dev);
+	if (IS_ERR_OR_NULL(adpt->pinctrl)) {
+		netdev_err(adpt->netdev, "error:%ld Failed to get pin ctrl\n",
+				PTR_ERR(adpt->pinctrl));
+		return PTR_ERR(adpt->pinctrl);
+	}
+	adpt->mdio_pins_active = pinctrl_lookup_state(adpt->pinctrl,
+				EMAC_PINCTRL_STATE_MDIO_ACTIVE);
+	if (IS_ERR_OR_NULL(adpt->mdio_pins_active)) {
+		netdev_err(adpt->netdev, "error:%ld Failed to lookup mdio pinctrl active state\n",
+				PTR_ERR(adpt->mdio_pins_active));
+		return PTR_ERR(adpt->mdio_pins_active);
+	}
+
+	adpt->mdio_pins_sleep = pinctrl_lookup_state(adpt->pinctrl,
+				EMAC_PINCTRL_STATE_MDIO_SLEEP);
+	if (IS_ERR_OR_NULL(adpt->mdio_pins_sleep)) {
+		netdev_err(adpt->netdev, "error:%ld Failed to lookup mdio pinctrl sleep state\n",
+				PTR_ERR(adpt->mdio_pins_sleep));
+		return PTR_ERR(adpt->mdio_pins_sleep);
+	}
+
+	adpt->ephy_pins_active = pinctrl_lookup_state(adpt->pinctrl,
+				EMAC_PINCTRL_STATE_EPHY_ACTIVE);
+	if (IS_ERR_OR_NULL(adpt->ephy_pins_active)) {
+		netdev_err(adpt->netdev, "error:%ld Failed to lookup ephy pinctrl active state\n",
+				PTR_ERR(adpt->ephy_pins_active));
+		return PTR_ERR(adpt->ephy_pins_active);
+	}
+
+	adpt->ephy_pins_sleep = pinctrl_lookup_state(adpt->pinctrl,
+				EMAC_PINCTRL_STATE_EPHY_SLEEP);
+	if (IS_ERR_OR_NULL(adpt->ephy_pins_sleep)) {
+		netdev_err(adpt->netdev, "error:%ld Failed to lookup ephy pinctrl sleep state\n",
+				PTR_ERR(adpt->ephy_pins_sleep));
+		return PTR_ERR(adpt->ephy_pins_sleep);
+	}
+
+	return 0;
+}
+
 /* Get the resources */
 static int emac_probe_resources(struct platform_device *pdev,
 				struct emac_adapter *adpt)
@@ -564,6 +672,14 @@ static int emac_probe_resources(struct p
 	else
 		eth_hw_addr_random(netdev);
 
+	/* Get pinctrl */
+	ret = msm_emac_pinctrl_init(adpt, &pdev->dev);
+	if (ret) 
+		return ret;
+
+	adpt->gpio_on = msm_emac_request_pinctrl_on;
+	adpt->gpio_off = msm_emac_request_pinctrl_off;
+
 	/* Core 0 interrupt */
 	ret = platform_get_irq(pdev, 0);
 	if (ret < 0)
@@ -606,6 +722,104 @@ static const struct acpi_device_id emac_
 MODULE_DEVICE_TABLE(acpi, emac_acpi_match);
 #endif
 
+/* Get the regulator */
+static int emac_get_regulator(struct platform_device *pdev,
+			      struct emac_adapter *adpt)
+{
+	struct regulator *vreg;
+	u8 i;
+	int len = 0;
+	u32 tmp[EMAC_VREG_CNT];
+
+	for (i = 0; i < EMAC_VREG_CNT; i++) {
+		vreg = devm_regulator_get(&pdev->dev, emac_regulator_name[i]);
+
+		if (IS_ERR(vreg)) {
+			netdev_err(adpt->netdev, "error:%ld unable to get emac %s\n",
+				 PTR_ERR(vreg), emac_regulator_name[i]);
+			return PTR_ERR(vreg);
+		}
+		adpt->vreg[i].vreg = vreg;
+	}
+
+	if (of_get_property(pdev->dev.of_node,
+			    "qcom,vdd-voltage-level", &len)) {
+		if (len == sizeof(tmp)) {
+			of_property_read_u32_array(pdev->dev.of_node,
+						   "qcom,vdd-voltage-level",
+						   tmp, len/sizeof(*tmp));
+
+			for (i = 0; i < EMAC_VREG_CNT; i++)
+				adpt->vreg[i].voltage_uv = tmp[i];
+		} else {
+			netdev_err(adpt->netdev, "unable to read voltage values for emac LDOs\n");
+			return -EINVAL;
+		}
+	} else {
+		netdev_err(adpt->netdev, "unable to read qcom,vdd-voltage-level emac dt property\n");
+		return -EINVAL;
+	}
+	return 0;
+}
+
+/* Set the Voltage */
+static int emac_set_voltage(struct emac_adapter *adpt, enum emac_vreg_id id,
+			    int min_uV, int max_uV)
+{
+	int retval = regulator_set_voltage(adpt->vreg[id].vreg, min_uV, max_uV);
+	if (retval)
+		netdev_err(adpt->netdev, "error:%d set voltage for %s\n",
+			 retval, emac_regulator_name[id]);
+	return retval;
+}
+
+/* Enable the emac core, internal/external phy regulator */
+static int emac_enable_regulator(struct emac_adapter *adpt, u8 start, u8 end)
+{
+	int retval = 0;
+	u8 i;
+
+	for (i = start; i <= end; i++) {
+		if (adpt->vreg[i].enabled)
+			continue;
+
+		if (adpt->vreg[i].voltage_uv) {
+			retval = emac_set_voltage(adpt, i,
+						  adpt->vreg[i].voltage_uv,
+						  adpt->vreg[i].voltage_uv);
+			if (retval)
+				goto err;
+		}
+
+		retval = regulator_enable(adpt->vreg[i].vreg);
+		if (retval) {
+			netdev_err(adpt->netdev, "error:%d enable regulator %s\n",
+				 retval, emac_regulator_name[EMAC_VREG3]);
+			goto err;
+		} else {
+			adpt->vreg[i].enabled = true;
+		}
+	}
+err:
+	return retval;
+}
+
+/* LDO init */
+static int msm_emac_ldo_init(struct platform_device *pdev,
+			     struct emac_adapter *adpt)
+{
+	int retval = 0;
+
+	retval = emac_get_regulator(pdev, adpt);
+	if (retval)
+		return retval;
+
+	retval =  emac_enable_regulator(adpt, EMAC_VREG1, EMAC_VREG5);
+	if (retval)
+		return retval;
+	return 0;
+}
+
 static int emac_probe(struct platform_device *pdev)
 {
 	struct net_device *netdev;
@@ -649,6 +863,10 @@ static int emac_probe(struct platform_de
 	if (ret)
 		goto err_undo_netdev;
 
+	ret = msm_emac_ldo_init(pdev, adpt);
+	if (ret)
+		goto err_undo_netdev;
+
 	/* initialize clocks */
 	ret = emac_clks_phase1_init(pdev, adpt);
 	if (ret) {
@@ -680,6 +898,11 @@ static int emac_probe(struct platform_de
 		goto err_undo_mdiobus;
 	}
 
+	/* Configure MDIO lines */
+	ret = adpt->gpio_on(adpt, true, true);
+	if (ret)
+		goto err_undo_mdiobus;
+
 	/* set hw features */
 	netdev->features = NETIF_F_SG | NETIF_F_HW_CSUM | NETIF_F_RXCSUM |
 			NETIF_F_TSO | NETIF_F_TSO6 | NETIF_F_HW_VLAN_CTAG_RX |
--- a/drivers/net/ethernet/qualcomm/emac/emac.h
+++ b/drivers/net/ethernet/qualcomm/emac/emac.h
@@ -200,6 +200,15 @@ enum emac_clk_id {
 	EMAC_CLK_CNT
 };
 
+enum emac_vreg_id {
+	EMAC_VREG1,
+	EMAC_VREG2,
+	EMAC_VREG3,
+	EMAC_VREG4,
+	EMAC_VREG5,
+	EMAC_VREG_CNT
+};
+
 #define EMAC_LINK_SPEED_UNKNOWN                                    0x0
 #define EMAC_LINK_SPEED_10_HALF                                 BIT(0)
 #define EMAC_LINK_SPEED_10_FULL                                 BIT(1)
@@ -321,6 +330,12 @@ struct emac_irq {
 	u32		mask;
 };
 
+struct emac_regulator {
+	struct regulator *vreg;
+	int	voltage_uv;
+	bool	enabled;
+};
+
 /* The device's main data structure */
 struct emac_adapter {
 	struct net_device		*netdev;
@@ -335,6 +350,7 @@ struct emac_adapter {
 
 	struct emac_irq			irq;
 	struct clk			*clk[EMAC_CLK_CNT];
+	struct emac_regulator		vreg[EMAC_VREG_CNT];
 
 	/* All Descriptor memory */
 	struct emac_ring_header		ring_header;
@@ -375,6 +391,14 @@ struct emac_adapter {
 	u16				msg_enable;
 
 	struct mutex			reset_lock;
+
+	struct pinctrl			*pinctrl;
+	struct pinctrl_state		*mdio_pins_active;
+	struct pinctrl_state		*mdio_pins_sleep;
+	struct pinctrl_state		*ephy_pins_active;
+	struct pinctrl_state		*ephy_pins_sleep;
+	int	(*gpio_on)(struct emac_adapter *adpt, bool mdio, bool ephy);
+	int	(*gpio_off)(struct emac_adapter *adpt, bool mdio, bool ephy);
 };
 
 int emac_reinit_locked(struct emac_adapter *adpt);
