--- a/include/linux/netdevice.h
+++ b/include/linux/netdevice.h
@@ -2610,6 +2610,7 @@ enum netdev_cmd {
 	NETDEV_CVLAN_FILTER_DROP_INFO,
 	NETDEV_SVLAN_FILTER_PUSH_INFO,
 	NETDEV_SVLAN_FILTER_DROP_INFO,
+	NETDEV_UNREGISTER_FINAL,
 };
 const char *netdev_cmd_to_name(enum netdev_cmd cmd);
 
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -9515,6 +9515,8 @@ static void netdev_wait_allrefs(struct n
 			rcu_barrier();
 			rtnl_lock();
 
+			call_netdevice_notifiers(NETDEV_UNREGISTER_FINAL, dev);
+
 			if (test_bit(__LINK_STATE_LINKWATCH_PENDING,
 				     &dev->state)) {
 				/* We must not have linkwatch events
@@ -9586,6 +9588,10 @@ void netdev_run_todo(void)
 			= list_first_entry(&list, struct net_device, todo_list);
 		list_del(&dev->todo_list);
 
+                rtnl_lock();
+                call_netdevice_notifiers(NETDEV_UNREGISTER_FINAL, dev);
+                __rtnl_unlock();
+
 		if (unlikely(dev->reg_state != NETREG_UNREGISTERING)) {
 			pr_err("network todo '%s' but state %d\n",
 			       dev->name, dev->reg_state);
@@ -10023,6 +10029,7 @@ int dev_change_net_namespace(struct net_
 	 */
 	call_netdevice_notifiers(NETDEV_UNREGISTER, dev);
 	rcu_barrier();
+	call_netdevice_notifiers(NETDEV_UNREGISTER_FINAL, dev);
 
 	new_nsid = peernet2id_alloc(dev_net(dev), net, GFP_KERNEL);
 	/* If there is an ifindex conflict assign a new one */
