--- a/drivers/net/ethernet/qualcomm/emac/emac-mac.c
+++ b/drivers/net/ethernet/qualcomm/emac/emac-mac.c
@@ -446,14 +446,14 @@ static void emac_mac_config(struct emac_
 
 	emac_set_mac_address(adpt, netdev->dev_addr);
 
-	max_frame = netdev->mtu + ETH_HLEN + ETH_FCS_LEN + VLAN_HLEN;
+	// 2x VLAN_HLEN for double VLAN tagging
+	max_frame = netdev->mtu + ETH_HLEN + ETH_FCS_LEN + VLAN_HLEN * 2;
 	adpt->rxbuf_size = netdev->mtu > EMAC_DEF_RX_BUF_SIZE ?
 		ALIGN(max_frame, 8) : EMAC_DEF_RX_BUF_SIZE;
 
 	emac_mac_dma_rings_config(adpt);
 
-	writel(netdev->mtu + ETH_HLEN + VLAN_HLEN + ETH_FCS_LEN,
-	       adpt->base + EMAC_MAX_FRAM_LEN_CTRL);
+	writel(max_frame, adpt->base + EMAC_MAX_FRAM_LEN_CTRL);
 
 	emac_mac_tx_config(adpt);
 	emac_mac_rx_config(adpt);
@@ -1073,13 +1073,6 @@ static void emac_receive_skb(struct emac
 			     struct sk_buff *skb,
 			     u16 vlan_tag, bool vlan_flag)
 {
-	if (vlan_flag) {
-		u16 vlan;
-
-		EMAC_TAG_TO_VLAN(vlan_tag, vlan);
-		__vlan_hwaccel_put_tag(skb, htons(ETH_P_8021Q), vlan);
-	}
-
 	napi_gro_receive(&rx_q->napi, skb);
 }
 
--- a/drivers/net/ethernet/qualcomm/emac/emac.c
+++ b/drivers/net/ethernet/qualcomm/emac/emac.c
@@ -937,18 +937,17 @@ static int emac_probe(struct platform_de
 
 	/* set hw features */
 	netdev->features = NETIF_F_SG | NETIF_F_HW_CSUM | NETIF_F_RXCSUM |
-			NETIF_F_TSO | NETIF_F_TSO6 | NETIF_F_HW_VLAN_CTAG_RX |
-			NETIF_F_HW_VLAN_CTAG_TX;
+			NETIF_F_TSO | NETIF_F_TSO6;
 	netdev->hw_features = netdev->features;
 
 	netdev->vlan_features |= NETIF_F_SG | NETIF_F_HW_CSUM |
 				 NETIF_F_TSO | NETIF_F_TSO6;
 
-	/* MTU range: 46 - 9194 */
+	/* MTU range: 42 - 9190 */
 	netdev->min_mtu = EMAC_MIN_ETH_FRAME_SIZE -
-			  (ETH_HLEN + ETH_FCS_LEN + VLAN_HLEN);
+			  (ETH_HLEN + ETH_FCS_LEN + VLAN_HLEN * 2);
 	netdev->max_mtu = EMAC_MAX_ETH_FRAME_SIZE -
-			  (ETH_HLEN + ETH_FCS_LEN + VLAN_HLEN);
+			  (ETH_HLEN + ETH_FCS_LEN + VLAN_HLEN * 2);
 
 	INIT_WORK(&adpt->work_thread, emac_work_thread);
 
