--- a/drivers/soc/qcom/smp2p.c
+++ b/drivers/soc/qcom/smp2p.c
@@ -1,7 +1,16 @@
 // SPDX-License-Identifier: GPL-2.0-only
 /*
  * Copyright (c) 2015, Sony Mobile Communications AB.
- * Copyright (c) 2012-2013, The Linux Foundation. All rights reserved.
+ * Copyright (c) 2012-2013, 2018-2020 The Linux Foundation. All rights reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 and
+ * only version 2 as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
  */
 
 #include <linux/interrupt.h>
@@ -18,6 +27,9 @@
 #include <linux/soc/qcom/smem.h>
 #include <linux/soc/qcom/smem_state.h>
 #include <linux/spinlock.h>
+#include <linux/pm_wakeup.h>
+
+#include <linux/ipc_logging.h>
 
 /*
  * The Shared Memory Point to Point (SMP2P) protocol facilitates communication
@@ -40,8 +52,12 @@
 #define SMP2P_MAX_ENTRY_NAME 16
 
 #define SMP2P_FEATURE_SSR_ACK 0x1
+#define SMP2P_FLAGS_RESTART_DONE_BIT 0
+#define SMP2P_FLAGS_RESTART_ACK_BIT 1
 
 #define SMP2P_MAGIC 0x504d5324
+#define SMP2P_VERSION	1
+#define SMP2P_FEATURES	SMP2P_FEATURE_SSR_ACK
 
 /**
  * struct smp2p_smem_item - in memory communication structure
@@ -97,6 +113,7 @@ struct smp2p_entry {
 
 	struct irq_domain *domain;
 	DECLARE_BITMAP(irq_enabled, 32);
+	DECLARE_BITMAP(irq_pending, 32);
 	DECLARE_BITMAP(irq_rising, 32);
 	DECLARE_BITMAP(irq_falling, 32);
 
@@ -134,12 +151,18 @@ struct qcom_smp2p {
 
 	unsigned valid_entries;
 
+	bool ssr_ack_enabled;
+	bool ssr_ack;
+	bool open;
+
 	unsigned local_pid;
 	unsigned remote_pid;
 
+	int irq;
 	struct regmap *ipc_regmap;
 	int ipc_offset;
 	int ipc_bit;
+	struct wakeup_source *ws;
 
 	struct mbox_client mbox_client;
 	struct mbox_chan *mbox_chan;
@@ -148,6 +171,14 @@ struct qcom_smp2p {
 	struct list_head outbound;
 };
 
+static void *ilc;
+#define SMP2P_LOG_PAGE_CNT 2
+#define SMP2P_INFO(x, ...)	\
+do {	\
+	if (ilc) \
+		ipc_log_string(ilc, "[%s]: "x, __func__, ##__VA_ARGS__); \
+} while (0)
+
 static void qcom_smp2p_kick(struct qcom_smp2p *smp2p)
 {
 	/* Make sure any updated data is written before the kick */
@@ -161,41 +192,66 @@ static void qcom_smp2p_kick(struct qcom_
 	}
 }
 
-/**
- * qcom_smp2p_intr() - interrupt handler for incoming notifications
- * @irq:	unused
- * @data:	smp2p driver context
- *
- * Handle notifications from the remote side to handle newly allocated entries
- * or any changes to the state bits of existing entries.
- */
-static irqreturn_t qcom_smp2p_intr(int irq, void *data)
+static bool qcom_smp2p_check_ssr(struct qcom_smp2p *smp2p)
 {
-	struct smp2p_smem_item *in;
-	struct smp2p_entry *entry;
-	struct qcom_smp2p *smp2p = data;
-	unsigned smem_id = smp2p->smem_items[SMP2P_INBOUND];
-	unsigned pid = smp2p->remote_pid;
-	size_t size;
-	int irq_pin;
-	u32 status;
-	char buf[SMP2P_MAX_ENTRY_NAME];
+	struct smp2p_smem_item *in = smp2p->in;
+	bool restart;
+
+	if (!smp2p->ssr_ack_enabled)
+		return false;
+
+	restart = in->flags & BIT(SMP2P_FLAGS_RESTART_DONE_BIT);
+	if (restart == smp2p->ssr_ack)
+		return false;
+
+	SMP2P_INFO("%d: SSR DETECTED\n", smp2p->remote_pid);
+	return true;
+}
+
+static void qcom_smp2p_do_ssr_ack(struct qcom_smp2p *smp2p)
+{
+	struct smp2p_smem_item *out = smp2p->out;
+	u32 ack = !smp2p->ssr_ack;
 	u32 val;
-	int i;
 
-	in = smp2p->in;
+	smp2p->ssr_ack = ack;
+	ack = ack << SMP2P_FLAGS_RESTART_ACK_BIT;
 
-	/* Acquire smem item, if not already found */
-	if (!in) {
-		in = qcom_smem_get(pid, smem_id, &size);
-		if (IS_ERR(in)) {
-			dev_err(smp2p->dev,
-				"Unable to acquire remote smp2p item\n");
-			return IRQ_HANDLED;
-		}
+	val = out->flags & ~BIT(SMP2P_FLAGS_RESTART_ACK_BIT);
+	val |= ack;
+	out->flags = val;
 
-		smp2p->in = in;
+	qcom_smp2p_kick(smp2p);
+}
+
+static void qcom_smp2p_negotiate(struct qcom_smp2p *smp2p)
+{
+	struct smp2p_smem_item *out = smp2p->out;
+	struct smp2p_smem_item *in = smp2p->in;
+	u32 features;
+
+	if (in->version == out->version) {
+		features = in->features & out->features;
+		out->features = features;
+
+		if (features & SMP2P_FEATURE_SSR_ACK)
+			smp2p->ssr_ack_enabled = true;
+
+		smp2p->open = true;
+		SMP2P_INFO("%d: state=open ssr_ack=%d\n", smp2p->remote_pid,
+			   smp2p->ssr_ack_enabled);
 	}
+}
+
+static void qcom_smp2p_notify_in(struct qcom_smp2p *smp2p)
+{
+	struct smp2p_smem_item *in = smp2p->in;
+	struct smp2p_entry *entry;
+	unsigned long status;
+	int irq_pin;
+	char buf[SMP2P_MAX_ENTRY_NAME];
+	u32 val;
+	int i;
 
 	/* Match newly created entries */
 	for (i = smp2p->valid_entries; i < in->valid_entries; i++) {
@@ -219,23 +275,79 @@ static irqreturn_t qcom_smp2p_intr(int i
 
 		status = val ^ entry->last_value;
 		entry->last_value = val;
+		status |= *entry->irq_pending;
 
 		/* No changes of this entry? */
 		if (!status)
 			continue;
 
-		for_each_set_bit(i, entry->irq_enabled, 32) {
-			if (!(status & BIT(i)))
-				continue;
+		SMP2P_INFO("%d: %s: status:%0lx val:%0x\n",
+			   smp2p->remote_pid, entry->name, status, val);
 
+		for_each_set_bit(i, &status, 32) {
 			if ((val & BIT(i) && test_bit(i, entry->irq_rising)) ||
 			    (!(val & BIT(i)) && test_bit(i, entry->irq_falling))) {
 				irq_pin = irq_find_mapping(entry->domain, i);
 				handle_nested_irq(irq_pin);
+				clear_bit(i, entry->irq_pending);
 			}
 		}
 	}
+}
+
+static irqreturn_t qcom_smp2p_isr(int irq, void *data)
+{
+	struct qcom_smp2p *smp2p = data;
+
+	__pm_stay_awake(smp2p->ws);
+	return IRQ_WAKE_THREAD;
+}
+
+/**
+ * qcom_smp2p_intr() - interrupt handler for incoming notifications
+ * @irq:	unused
+ * @data:	smp2p driver context
+ *
+ * Handle notifications from the remote side to handle newly allocated entries
+ * or any changes to the state bits of existing entries.
+ */
+static irqreturn_t qcom_smp2p_intr(int irq, void *data)
+{
+	struct smp2p_smem_item *in;
+	struct qcom_smp2p *smp2p = data;
+	unsigned int smem_id = smp2p->smem_items[SMP2P_INBOUND];
+	unsigned int pid = smp2p->remote_pid;
+	size_t size;
+
+	in = smp2p->in;
 
+	/* Acquire smem item, if not already found */
+	if (!in) {
+		in = qcom_smem_get(pid, smem_id, &size);
+		if (IS_ERR(in)) {
+			dev_err(smp2p->dev,
+				"Unable to acquire remote smp2p item\n");
+			goto out;
+		}
+
+		smp2p->in = in;
+	}
+
+	if (!smp2p->open)
+		qcom_smp2p_negotiate(smp2p);
+
+	if (smp2p->open) {
+		bool do_restart;
+
+		do_restart = qcom_smp2p_check_ssr(smp2p);
+		qcom_smp2p_notify_in(smp2p);
+
+		if (do_restart)
+			qcom_smp2p_do_ssr_ack(smp2p);
+	}
+
+out:
+	__pm_relax(smp2p->ws);
 	return IRQ_HANDLED;
 }
 
@@ -276,11 +388,23 @@ static int smp2p_set_irq_type(struct irq
 	return 0;
 }
 
+static int smp2p_retrigger_irq(struct irq_data *irqd)
+{
+	struct smp2p_entry *entry = irq_data_get_irq_chip_data(irqd);
+	irq_hw_number_t irq = irqd_to_hwirq(irqd);
+
+	SMP2P_INFO("%d: %s: %lu\n", entry->smp2p->remote_pid, entry->name, irq);
+	set_bit(irq, entry->irq_pending);
+
+	return 0;
+}
+
 static struct irq_chip smp2p_irq_chip = {
 	.name           = "smp2p",
 	.irq_mask       = smp2p_mask_irq,
 	.irq_unmask     = smp2p_unmask_irq,
 	.irq_set_type	= smp2p_set_irq_type,
+	.irq_retrigger	= smp2p_retrigger_irq,
 };
 
 static int smp2p_irq_map(struct irq_domain *d,
@@ -293,6 +417,8 @@ static int smp2p_irq_map(struct irq_doma
 	irq_set_chip_data(irq, entry);
 	irq_set_nested_thread(irq, 1);
 	irq_set_noprobe(irq);
+	irq_set_parent(irq, entry->smp2p->irq);
+	irq_set_status_flags(irq, IRQ_DISABLE_UNLAZY);
 
 	return 0;
 }
@@ -318,16 +444,17 @@ static int qcom_smp2p_inbound_entry(stru
 static int smp2p_update_bits(void *data, u32 mask, u32 value)
 {
 	struct smp2p_entry *entry = data;
-	unsigned long flags;
 	u32 orig;
 	u32 val;
 
-	spin_lock_irqsave(&entry->lock, flags);
+	spin_lock(&entry->lock);
 	val = orig = readl(entry->value);
 	val &= ~mask;
 	val |= value;
 	writel(val, entry->value);
-	spin_unlock_irqrestore(&entry->lock, flags);
+	spin_unlock(&entry->lock);
+	SMP2P_INFO("%d: %s: orig:0x%0x new:0x%0x\n",
+		   entry->smp2p->remote_pid, entry->name, orig, val);
 
 	if (val != orig)
 		qcom_smp2p_kick(entry->smp2p);
@@ -391,13 +518,14 @@ static int qcom_smp2p_alloc_outbound_ite
 	out->remote_pid = smp2p->remote_pid;
 	out->total_entries = SMP2P_MAX_ENTRY;
 	out->valid_entries = 0;
+	out->features = SMP2P_FEATURES;
 
 	/*
 	 * Make sure the rest of the header is written before we validate the
 	 * item by writing a valid version number.
 	 */
 	wmb();
-	out->version = 1;
+	out->version = SMP2P_VERSION;
 
 	qcom_smp2p_kick(smp2p);
 
@@ -440,15 +568,63 @@ static int smp2p_parse_ipc(struct qcom_s
 	return 0;
 }
 
-static int qcom_smp2p_probe(struct platform_device *pdev)
+static int qcom_smp2p_alloc_item(struct platform_device *pdev,
+					struct qcom_smp2p *smp2p)
 {
+	int ret = 0;
 	struct smp2p_entry *entry;
 	struct device_node *node;
+
+	ret = qcom_smp2p_alloc_outbound_item(smp2p);
+	if (ret < 0)
+		return ret;
+
+	for_each_available_child_of_node(pdev->dev.of_node, node) {
+		entry = devm_kzalloc(&pdev->dev, sizeof(*entry), GFP_KERNEL);
+		if (!entry)
+			return -ENOMEM;
+
+		entry->smp2p = smp2p;
+		spin_lock_init(&entry->lock);
+
+		ret = of_property_read_string(node, "qcom,entry-name",
+							&entry->name);
+		if (ret < 0)
+			return ret;
+
+		if (of_property_read_bool(node, "interrupt-controller")) {
+			ret = qcom_smp2p_inbound_entry(smp2p, entry, node);
+			if (ret < 0)
+				return ret;
+
+			list_add(&entry->node, &smp2p->inbound);
+		} else {
+			ret = qcom_smp2p_outbound_entry(smp2p, entry, node);
+			if (ret < 0)
+				return ret;
+
+			list_add(&entry->node, &smp2p->outbound);
+		}
+	}
+	smp2p->ws = wakeup_source_create("smp2p");
+	wakeup_source_add(smp2p->ws);
+
+	/* Kick the outgoing edge after allocating entries */
+	qcom_smp2p_kick(smp2p);
+
+	return ret;
+}
+
+static int qcom_smp2p_probe(struct platform_device *pdev)
+{
+	struct smp2p_entry *entry;
 	struct qcom_smp2p *smp2p;
 	const char *key;
-	int irq;
 	int ret;
 
+	if (!ilc)
+		ilc = ipc_log_context_create(SMP2P_LOG_PAGE_CNT, "smp2p", 0);
+
 	smp2p = devm_kzalloc(&pdev->dev, sizeof(*smp2p), GFP_KERNEL);
 	if (!smp2p)
 		return -ENOMEM;
@@ -467,18 +643,22 @@ static int qcom_smp2p_probe(struct platf
 
 	key = "qcom,local-pid";
 	ret = of_property_read_u32(pdev->dev.of_node, key, &smp2p->local_pid);
-	if (ret)
-		goto report_read_failure;
+	if (ret < 0) {
+		dev_err(&pdev->dev, "failed to read %s\n", key);
+		return -EINVAL;
+	}
 
 	key = "qcom,remote-pid";
 	ret = of_property_read_u32(pdev->dev.of_node, key, &smp2p->remote_pid);
-	if (ret)
-		goto report_read_failure;
+	if (ret < 0) {
+		dev_err(&pdev->dev, "failed to read %s\n", key);
+		return -EINVAL;
+	}
 
-	irq = platform_get_irq(pdev, 0);
-	if (irq < 0) {
+	smp2p->irq = platform_get_irq(pdev, 0);
+	if (smp2p->irq < 0) {
 		dev_err(&pdev->dev, "unable to acquire smp2p interrupt\n");
-		return irq;
+		return smp2p->irq;
 	}
 
 	smp2p->mbox_client.dev = &pdev->dev;
@@ -496,51 +676,26 @@ static int qcom_smp2p_probe(struct platf
 			return ret;
 	}
 
-	ret = qcom_smp2p_alloc_outbound_item(smp2p);
-	if (ret < 0)
-		goto release_mbox;
-
-	for_each_available_child_of_node(pdev->dev.of_node, node) {
-		entry = devm_kzalloc(&pdev->dev, sizeof(*entry), GFP_KERNEL);
-		if (!entry) {
-			ret = -ENOMEM;
-			goto unwind_interfaces;
-		}
-
-		entry->smp2p = smp2p;
-		spin_lock_init(&entry->lock);
-
-		ret = of_property_read_string(node, "qcom,entry-name", &entry->name);
-		if (ret < 0)
-			goto unwind_interfaces;
-
-		if (of_property_read_bool(node, "interrupt-controller")) {
-			ret = qcom_smp2p_inbound_entry(smp2p, entry, node);
-			if (ret < 0)
-				goto unwind_interfaces;
-
-			list_add(&entry->node, &smp2p->inbound);
-		} else  {
-			ret = qcom_smp2p_outbound_entry(smp2p, entry, node);
-			if (ret < 0)
-				goto unwind_interfaces;
-
-			list_add(&entry->node, &smp2p->outbound);
+	ret = qcom_smp2p_alloc_item(pdev, smp2p);
+	if (ret < 0 && ret != -EEXIST) {
+		if (ret != -EPROBE_DEFER) {
+			dev_err(&pdev->dev, "failed to alloc outbound items\n");
+			goto release_mbox;
 		}
+	} else if (ret) {
+		dev_err(&pdev->dev, "failed to get smp2p entries\n");
+		goto unwind_interfaces;
 	}
 
-	/* Kick the outgoing edge after allocating entries */
-	qcom_smp2p_kick(smp2p);
-
-	ret = devm_request_threaded_irq(&pdev->dev, irq,
-					NULL, qcom_smp2p_intr,
-					IRQF_ONESHOT,
+	ret = devm_request_threaded_irq(&pdev->dev, smp2p->irq,
+					qcom_smp2p_isr, qcom_smp2p_intr,
+					IRQF_NO_SUSPEND | IRQF_ONESHOT,
 					"smp2p", (void *)smp2p);
 	if (ret) {
 		dev_err(&pdev->dev, "failed to request interrupt\n");
 		goto unwind_interfaces;
 	}
-
+	enable_irq_wake(smp2p->irq);
 
 	return 0;
 
@@ -557,10 +712,6 @@ release_mbox:
 	mbox_free_channel(smp2p->mbox_chan);
 
 	return ret;
-
-report_read_failure:
-	dev_err(&pdev->dev, "failed to read %s\n", key);
-	return -EINVAL;
 }
 
 static int qcom_smp2p_remove(struct platform_device *pdev)
@@ -581,6 +732,89 @@ static int qcom_smp2p_remove(struct plat
 	return 0;
 }
 
+static int qcom_smp2p_restore(struct device *dev)
+{
+	int ret = 0;
+	struct qcom_smp2p *smp2p = dev_get_drvdata(dev);
+	struct smp2p_entry *entry;
+	struct device_node *node;
+	struct platform_device *pdev = container_of(dev, struct
+						platform_device, dev);
+
+	ret = qcom_smp2p_alloc_outbound_item(smp2p);
+	if (ret < 0)
+		goto print_err;
+
+	for_each_available_child_of_node(pdev->dev.of_node, node) {
+		entry = devm_kzalloc(&pdev->dev, sizeof(*entry), GFP_KERNEL);
+		if (!entry) {
+			ret = -ENOMEM;
+			goto print_err;
+		}
+
+		entry->smp2p = smp2p;
+		spin_lock_init(&entry->lock);
+		ret = of_property_read_string(node, "qcom,entry-name",
+								&entry->name);
+		if (ret < 0)
+			goto print_err;
+
+		if (!of_property_read_bool(node, "interrupt-controller")) {
+			ret = qcom_smp2p_outbound_entry(smp2p, entry, node);
+			if (ret < 0)
+				goto print_err;
+
+			list_add(&entry->node, &smp2p->outbound);
+		}
+	}
+	smp2p->ws = wakeup_source_create("smp2p");
+	wakeup_source_add(smp2p->ws);
+
+	enable_irq_wake(smp2p->irq);
+	/* Kick the outgoing edge after allocating entries */
+	qcom_smp2p_kick(smp2p);
+
+print_err:
+	if (ret < 0 && ret != -EEXIST)
+		dev_err(dev, "failed to alloc items ret = %d\n", ret);
+
+	return ret;
+}
+
+static int qcom_smp2p_freeze(struct device *dev)
+{
+	struct qcom_smp2p *smp2p = dev_get_drvdata(dev);
+	struct smp2p_entry *entry;
+	struct smp2p_entry *next_entry;
+
+	disable_irq_wake(smp2p->irq);
+	/* Walk through the out bound list and release state and entry */
+	list_for_each_entry_safe(entry, next_entry, &smp2p->outbound, node) {
+		qcom_smem_state_unregister(entry->state);
+		list_del(&entry->node);
+		devm_kfree(smp2p->dev, entry);
+	}
+	INIT_LIST_HEAD(&smp2p->outbound);
+
+	/* Walk through the in bound list and reset last value */
+	list_for_each_entry_safe(entry, next_entry, &smp2p->inbound, node) {
+		entry->last_value = 0;
+	}
+	/* make null to point it to valid smem item during first interrupt */
+	smp2p->in = NULL;
+	smp2p->valid_entries = 0;
+	/* remove wakeup source */
+	wakeup_source_remove(smp2p->ws);
+	wakeup_source_destroy(smp2p->ws);
+	return 0;
+}
+
+static const struct dev_pm_ops qcom_smp2p_pm_ops = {
+	.freeze = qcom_smp2p_freeze,
+	.restore = qcom_smp2p_restore,
+	.thaw = qcom_smp2p_restore,
+};
+
 static const struct of_device_id qcom_smp2p_of_match[] = {
 	{ .compatible = "qcom,smp2p" },
 	{}
@@ -593,6 +827,7 @@ static struct platform_driver qcom_smp2p
 	.driver  = {
 		.name  = "qcom_smp2p",
 		.of_match_table = qcom_smp2p_of_match,
+		.pm = &qcom_smp2p_pm_ops,
 	},
 };
 module_platform_driver(qcom_smp2p_driver);
