--- a/drivers/spi/spidev.c
+++ b/drivers/spi/spidev.c
@@ -85,6 +85,24 @@ static unsigned bufsiz = 4096;
 module_param(bufsiz, uint, S_IRUGO);
 MODULE_PARM_DESC(bufsiz, "data bytes in biggest supported SPI message");
 
+static int busnum = 6;
+module_param(busnum, int, S_IRUGO);
+MODULE_PARM_DESC(busnum, "bus num of the controller");
+
+static int chipselect = 0;
+module_param(chipselect, int, S_IRUGO);
+MODULE_PARM_DESC(chipselect, "chip select of the desired device");
+
+static int maxspeed = 19200000;
+module_param(maxspeed, int, S_IRUGO);
+MODULE_PARM_DESC(maxspeed, "max_speed of the desired device");
+
+static int spimode = SPI_MODE_3;
+module_param(spimode, int, S_IRUGO);
+MODULE_PARM_DESC(spimode, "mode of the desired device");
+
+static struct spi_device *spi;
+
 /*-------------------------------------------------------------------------*/
 
 static ssize_t
@@ -857,12 +875,53 @@ static int __init spidev_init(void)
 		class_destroy(spidev_class);
 		unregister_chrdev(SPIDEV_MAJOR, spidev_spi_driver.driver.name);
 	}
+
+        if (busnum != -1 && chipselect != -1) {
+                struct spi_board_info chip = {
+                                        .modalias       = "spidev",
+                                        .mode           = spimode,
+                                        .bus_num        = busnum,
+                                        .chip_select    = chipselect,
+                                        .max_speed_hz   = maxspeed,
+                };
+
+                struct spi_master *master;
+
+                master = spi_busnum_to_master(busnum);  
+                if (!master) {
+                        status = -ENODEV;
+                        goto error_busnum;
+                }
+
+                /* We create a virtual device that will sit on the bus */
+                spi = spi_new_device(master, &chip);
+                if (!spi) {
+                        status = -EBUSY;
+                        goto error_mem;
+                }
+                dev_dbg(&spi->dev, "busnum=%d cs=%d bufsiz=%d maxspeed=%d",
+                        busnum, chipselect, bufsiz, maxspeed);
+        }
+
 	return status;
+
+error_mem:
+error_busnum:
+        spi_unregister_driver(&spidev_spi_driver);
+        class_destroy(spidev_class);
+        unregister_chrdev(SPIDEV_MAJOR, spidev_spi_driver.driver.name);
+
+        return status;
 }
 module_init(spidev_init);
 
 static void __exit spidev_exit(void)
 {
+        if (spi) {
+                spi_unregister_device(spi);
+                spi = NULL;
+        }
+
 	spi_unregister_driver(&spidev_spi_driver);
 	class_destroy(spidev_class);
 	unregister_chrdev(SPIDEV_MAJOR, spidev_spi_driver.driver.name);
