--- a/drivers/rtc/Makefile
+++ b/drivers/rtc/Makefile
@@ -134,6 +134,7 @@ obj-$(CONFIG_RTC_DRV_PM8XXX)	+= rtc-pm8x
 obj-$(CONFIG_RTC_DRV_PS3)	+= rtc-ps3.o
 obj-$(CONFIG_RTC_DRV_PUV3)	+= rtc-puv3.o
 obj-$(CONFIG_RTC_DRV_PXA)	+= rtc-pxa.o
+obj-$(CONFIG_RTC_DRV_QPNP)      += qpnp-rtc.o
 obj-$(CONFIG_RTC_DRV_R7301)	+= rtc-r7301.o
 obj-$(CONFIG_RTC_DRV_R9701)	+= rtc-r9701.o
 obj-$(CONFIG_RTC_DRV_RC5T583)	+= rtc-rc5t583.o
--- a/drivers/rtc/Kconfig
+++ b/drivers/rtc/Kconfig
@@ -1538,6 +1538,15 @@ config RTC_DRV_PXA
 	 This RTC driver uses PXA RTC registers available since pxa27x
 	 series (RDxR, RYxR) instead of legacy RCNR, RTAR.
 
+config RTC_DRV_QPNP
+        tristate "Qualcomm Technologies, Inc. QPNP PMIC RTC"
+        depends on SPMI
+        help
+          This enables support for the RTC found on Qualcomm Technologies, Inc.
+          QPNP PMIC chips.  This driver supports using the PMIC RTC peripheral
+          to wake a mobile device up from suspend or to wake it up from power-
+          off.
+
 config RTC_DRV_VT8500
 	tristate "VIA/WonderMedia 85xx SoC RTC"
 	depends on ARCH_VT8500 || COMPILE_TEST
