--- a/kernel/trace/Makefile
+++ b/kernel/trace/Makefile
@@ -85,4 +85,9 @@ obj-$(CONFIG_UPROBE_EVENTS) += trace_upr
 
 obj-$(CONFIG_TRACEPOINT_BENCHMARK) += trace_benchmark.o
 
+obj-$(CONFIG_IPC_LOGGING) += ipc_logging.o
+ifdef CONFIG_DEBUG_FS
+obj-$(CONFIG_IPC_LOGGING) += ipc_logging_debug.o
+endif
+
 libftrace-y := ftrace.o
--- a/kernel/trace/Kconfig
+++ b/kernel/trace/Kconfig
@@ -88,6 +88,17 @@ config RING_BUFFER_ALLOW_SWAP
 	 Allow the use of ring_buffer_swap_cpu.
 	 Adds a very slight overhead to tracing when enabled.
 
+config IPC_LOGGING
+        bool "Debug Logging for IPC Drivers"
+        select GENERIC_TRACER
+        help
+          IPC Logging driver provides a logging option for IPC Drivers.
+          This provides a cyclic buffer based logging support in a driver
+          specific context. This driver also provides a debugfs interface
+          to dump the logs in a live fashion.
+          
+          If in doubt, say no.
+
 config PREEMPTIRQ_TRACEPOINTS
 	bool
 	depends on TRACE_PREEMPT_TOGGLE || TRACE_IRQFLAGS
