--- a/drivers/mtd/nand/raw/nand_ids.c
+++ b/drivers/mtd/nand/raw/nand_ids.c
@@ -58,6 +58,26 @@ struct nand_flash_dev nand_flash_ids[] =
                   SZ_2K, SZ_512, SZ_128K, 0, 4, 128,
 		  NAND_ECC_INFO(4, SZ_512) },
 
+	{"MT29F4G08ABBDAHC 4G 3.3V 8-bit",
+		{ .id = {0x2c, 0xac, 0x90, 0x15, 0x00, 0x00, 0x00, 0x00} },
+		  SZ_2K, SZ_512, SZ_128K, 0, 4, 64,
+		  NAND_ECC_INFO(4, SZ_512) },
+
+	{"JSFCBB3Y7ABBD_2G 1.8V 8-bit",
+		{ .id = {0x01, 0xaC, 0x90, 0x15, 0x00, 0x00, 0x00, 0x00} },
+		  SZ_2K, SZ_512, SZ_128K, 0, 4, 128,
+		  NAND_ECC_INFO(8, SZ_512) },
+
+	{"JSFBB3YH3BBG 2G 1.8V 4-bit",
+		{ .id = {0xad, 0xaa, 0x90, 0x15, 0x00, 0x00, 0x00, 0x00} },
+		  SZ_2K, SZ_256, SZ_128K, 0, 4, 128,
+		  NAND_ECC_INFO(4, SZ_512) },
+
+	{"NM1282KSLAXAL 2G 1.8V 4-bit",
+		{ .id = {0xad, 0xac, 0x90, 0x15, 0x00, 0x00, 0x00, 0x00} },
+		  SZ_2K, SZ_512, SZ_128K, 0, 4, 128,
+		  NAND_ECC_INFO(4, SZ_512) },
+
 	LEGACY_ID_NAND("NAND 4MiB 5V 8-bit",   0x6B, 4, SZ_8K, SP_OPTIONS),
 	LEGACY_ID_NAND("NAND 4MiB 3,3V 8-bit", 0xE3, 4, SZ_8K, SP_OPTIONS),
 	LEGACY_ID_NAND("NAND 4MiB 3,3V 8-bit", 0xE5, 4, SZ_8K, SP_OPTIONS),
