--- a/mm/memblock.c
+++ b/mm/memblock.c
@@ -1755,6 +1755,14 @@ bool __init_memblock memblock_is_region_
 		 memblock.memory.regions[idx].size) >= end;
 }
 
+bool __init_memblock memblock_overlaps_memory(phys_addr_t base,
+                                              phys_addr_t size)
+{
+        memblock_cap_size(base, &size);
+
+        return memblock_overlaps_region(&memblock.memory, base, size);
+}
+
 /**
  * memblock_is_region_reserved - check if a region intersects reserved memory
  * @base: base of region to check
--- a/include/linux/memblock.h
+++ b/include/linux/memblock.h
@@ -449,6 +449,7 @@ bool memblock_is_map_memory(phys_addr_t
 bool memblock_is_region_memory(phys_addr_t base, phys_addr_t size);
 bool memblock_is_reserved(phys_addr_t addr);
 bool memblock_is_region_reserved(phys_addr_t base, phys_addr_t size);
+bool memblock_overlaps_memory(phys_addr_t base, phys_addr_t size);
 
 extern void __memblock_dump_all(void);
 
