--- a/drivers/soc/qcom/subsystem_restart.c
+++ b/drivers/soc/qcom/subsystem_restart.c
@@ -1215,9 +1215,10 @@ static void __subsystem_restart_dev(stru
 		} else {
 			panic("Subsystem %s crashed during SSR!", name);
 		}
-	} else
-		WARN(dev->track.state == SUBSYS_OFFLINE,
-			"SSR aborted: %s subsystem not online\n", name);
+	}
+	/*else*/
+		/*WARN(dev->track.state == SUBSYS_OFFLINE,*/
+			/*"SSR aborted: %s subsystem not online\n", name);*/
 	spin_unlock_irqrestore(&track->s_lock, flags);
 }
 
