Index: linux-msm5.4.244/drivers/pci/controller/pci-msm.c
===================================================================
--- linux-msm5.4.244.orig/drivers/pci/controller/pci-msm.c
+++ linux-msm5.4.244/drivers/pci/controller/pci-msm.c
@@ -6191,7 +6191,7 @@ static int msm_pcie_probe(struct platfor
 {
 	int ret = 0;
 	int rc_idx = -1;
-	int size;
+	int size = 0;
 	struct msm_pcie_dev_t *pcie_dev;
 	struct device_node *of_node;
 
